changes to MemIo::reserve() for #200

clanmills opened the pull request.

clanmills committed:
changes to MemIo::reserve() for #200

clanmills committed:
changes to MemIo::reserve() for #200

D4N commented on src/basicio.cpp:
```diff
@@ -1093,10 +1093,12 @@ namespace Exiv2 {
-            long size  = EXV_MAX(32768 * (1 + need / 32768), size_);
+            // Minimum size for 1st block
+            long size  = EXV_MAX(blockSize * (1 + need / blockSize), size_);
             byte* data = (byte*)std::malloc(size);
```
`malloc` can return `NULL` when the allocation request failed, we should check that here. Otherwise we'll get a segfault in the next line.

D4N commented on src/basicio.cpp:
```diff
@@ -1106,8 +1108,10 @@ namespace Exiv2 {
+                if ( blockSize > maxBlockSize ) blockSize = maxBlockSize ;
+                // Allocate in blocks
+                long want      = blockSize * (1 + need / blockSize );
                 data_ = (byte*)std::realloc(data_, want);
```
Same as `malloc`, can return `NULL`.

clanmills replied:
You're right.  I was so focused on the performance/blockSize matter that I didn't bother to think about malloc/realloc returning NULL.

I'm not so convinced about using std::vector.  MemIo is an in-memory file/stream in which the data is stored in a contiguous block of memory.  We could use std::vector to hold an array of bytes, however we probably need "blockSize" to request more bytes.  We could use std::vector to create an array of discontiguous blocks of memory.  This would avoid a single large contiguous memory block and complicate read() and write() to deal with block boundaries.  That design would avoid copying bytes in the file/stream as it grows.

When the bigtiff/multi-page tiff project moves forward, we could be working with medical imaging files of 10+Gbtyes from the folks at Imperial College.  It's unlikely that MemIo objects will be viable and we should re-instate the use of a temporary file on local storage in tiffimage.cpp/bigtiffimage.cpp

For the moment, I propose to:
1) Test/throw for NULL from malloc/realloc
2) retain the malloc/realloc contiguous block strategy

D4N commented:
Looks good to me.

I'd like to add, that we are basically reimplementing `std::vector` here, at least it's allocation strategy. So for the future we could use `std::vector` instead, as it handles all the memory management stuff for us.

clanmills committed:
Merge remote-tracking branch 'upstream/master'

clanmills commented:
@D4N Can you review this please.  I've added the malloc/NULL check and added a new error type in errors.cpp.  For the life of me, I have no idea what git/github does.  I see travis has started doing something, so I think the code is not somewhere on GitHub.

D4N commented:
The code is indeed not on Github, you have only pushed a merge commit. Have you commited the changes already? If yes, to which branch?

clanmills committed:
Changes following review by Dan (thanks, Dan).

clanmills committed:
Changes following review by Dan (thanks, Dan).

clanmills committed:
Merge branch 'clanmills-master' (#200 - changes following review by Dan).

D4N commented on src/basicio.cpp:
```diff
@@ -1100,6 +1100,9 @@ namespace Exiv2 {
             long size  = EXV_MAX(blockSize * (1 + need / blockSize), size_);
             byte* data = (byte*)std::malloc(size);
+            if (  data == NULL ) {
+                throw(60);
```
You are throwing an integer and not `Error(60)`.

D4N commented on src/basicio.cpp:
```diff
@@ -1093,11 +1093,16 @@ namespace Exiv2 {
+            long size  = EXV_MAX(blockSize * (1 + need / blockSize), size_);
             byte* data = (byte*)std::malloc(size);
+            if (  data == NULL ) {
+                throw(60);
```
You are throwing an integer here, I guess you want to do the following: `throw Error(60);`

clanmills replied:
Of course.  I'm so confused and baffled by Git that I've overlooked the obvious blunder.

It has taken hours to submit this change.  I can't understand why git/github is so complicated.

D4N replied:
I think to make this change you just have to switch to the branch where you made the previous change, correct the two lines, do a `git commit` and `git push`. You could also amend the changes with `git commit --amend` but that will require a `git push -f` (which can be scary, as you can overwrite history with that).

clanmills committed:
Trying again to submit code to issue #200

clanmills committed:
Merge branch 'Issue200'

# Conflicts:
#	src/basicio.cpp

D4N commented:
Looks good to me.

I would merge this once has Travis passed. I would also squash all your commits into one, if you don't mind, as the multiple merges make the history quite non-linear.

clanmills commented:
Thanks for doing this, Dan.  I was amazed when SmartGit posted 4 commits into the PR.  Astonished.  No idea why he did that.  I need a formula for a PR that I can follow without needing to understand git.  Once I've successfully submitted a few PRs, I'll try to understand how it works.
